epoch_publisher/src/client.rs

```diff
@@ -0,0 +1,149 @@
+use std::{collections::HashMap, sync::Arc};
+
+use crate::error::Error;
+use bytes::Bytes;
+use common::{host_info::HostInfo, network::fast_network::FastNetwork};
+use flatbuf::epoch_publisher_flatbuffers::epoch_publisher::*;
+use flatbuffers::FlatBufferBuilder;
+use tokio::sync::{oneshot, Mutex};
+use tokio_util::sync::CancellationToken;
+use uuid::Uuid;
+
+/// Provides an async rpc interface to a specific epoch publisher.
+pub struct EpochPublisherClient {
+    fast_network: Arc<dyn FastNetwork>,
+    range_server_info: HostInfo,
+    // TODO: make more typeful and store more information to e.g. allow timing out.
+    outstanding_requests: Mutex<HashMap<Uuid, oneshot::Sender<Bytes>>>,
+}
+
+pub fn serialize_uuid(uuid: Uuid) -> Uuidu128Args {
+    let uint128 = uuid.to_u128_le();
+    let lower = uint128 as u64;
+    let upper = (uint128 >> 64) as u64;
+    Uuidu128Args { lower, upper }
+}
+
+pub fn deserialize_uuid(uuidf: Uuidu128<'_>) -> Uuid {
+    let res: u128 = ((uuidf.upper() as u128) << 64) | (uuidf.lower() as u128);
+    Uuid::from_u128_le(res)
+}
+
+impl EpochPublisherClient {
+    pub fn new(
+        fast_network: Arc<dyn FastNetwork>,
+        runtime: tokio::runtime::Handle,
+        host_info: HostInfo,
+        cancellation_token: CancellationToken,
+    ) -> Arc<EpochPublisherClient> {
+        let pc = Arc::new(EpochPublisherClient {
+            fast_network,
+            range_server_info: host_info,
+            outstanding_requests: Mutex::new(HashMap::new()),
+        });
+
+        let pc_clone = pc.clone();
+        runtime.spawn(async move {
+            let _ = Self::network_loop(pc_clone, cancellation_token).await;
+            println!("Network loop exited!")
+        });
+        pc
+    }
+
+    pub async fn read_epoch(&self) -> Result<u64, Error> {
+        // TODO: gracefully handle malformed messages instead of unwrapping and crashing.
+        // TODO: too much copying :(
+        let req_id = Uuid::new_v4();
```

Comment:
Consider using a few sub routines here - for constructing the request, sending the req and parsing the response. That'd make the higher level logic more obvious. Some of these are probably generally useful. Feel free to do it as a follow-up.